kernel/src/ipc.rs

```diff
@@ -70,63 +80,67 @@ impl<const NUM_PROCS: usize> IPC<NUM_PROCS> {
                     IPCCallbackType::Client => match otherapp.index() {
-                        Some(i) => *mydata.client_callbacks.get(i).unwrap_or(&None),
-                        None => None,
+                        Some(i) => *mydata
+                            .client_callbacks
+                            .get(i)
+                            .unwrap_or(&Callback::default()),
+                        None => Callback::default(),
                     },
                 };
-                callback.map_or((), |mut callback| {
-                    self.data
-                        .enter(otherapp, |otherdata, _| {
-                            // If the other app shared a buffer with us, make
-                            // sure we have access to that slice and then call
-                            // the callback. If no slice was shared then just
-                            // call the callback.
-                            match appid.index() {
-                                Some(i) => {
-                                    if i >= otherdata.shared_memory.len() {
-                                        return;
-                                    }
+                self.data.enter(otherapp, |otherdata, _| {
+                    // If the other app shared a buffer with us, make
+                    // sure we have access to that slice and then call
+                    // the callback. If no slice was shared then just
+                    // call the callback.
+                    match appid.index() {
+                        Some(i) => {
+                            if i >= otherdata.shared_memory.len() {
+                                return;
+                            }
 
-                                    match otherdata.shared_memory[i] {
-                                        Some(ref slice) => {
-                                            slice.expose_to(appid);
-                                            callback.schedule(
-                                                otherapp.id() + 1,
-                                                slice.len(),
-                                                slice.ptr() as usize,
-                                            );
-                                        }
-                                        None => {
-                                            callback.schedule(otherapp.id() + 1, 0, 0);
-                                        }
-                                    }
+                            match otherdata.shared_memory.get(i) {
+                                Some(slice) => {
+                                    self.data.kernel.process_map_or(None, otherapp, |process| {
+                                        process.add_mpu_region(
+                                            slice.ptr(),
+                                            slice.len(),
+                                            slice.len(),
+                                        )
```

Comment:
We should probably look at https://github.com/tock/tock/issues/1993 since we are here. What happens if `add_mpu_region` fails?